### PR TITLE
Make `jvm_field_is` and `jvm_elem_is` check type compatibility.

### LINKED ISSUE #904: `jvm_field_is` and `jvm_elem_is` declarations should check types
Currently, when using `crucible_jvm_verify`, you can write specs that contain `jvm_field_is` or `jvm_elem_is` declarations with values of the wrong type. The type mismatch is not detected while these points-to declarations are being processed; however they may (or may not!) eventually lead to an "incorrect variant" error message later during symbolic simulation when that field or array element is read.

We should be checking for type compatibility of points-to declarations at an earlier phase, just like we do for LLVM specs. We also need to make some effort to ensure that specs used with `crucible_jvm_unsafe_assume_spec` are type-correct.

### PR DESCRIPTION
**Pull request by brianhuffman**
Fixes #904.

The check is performed at the spec-construction phase, as opposed
to during simulation; this means that type mismatches are caught
with either `crucible_jvm_verify` or `crucible_jvm_unsafe_assume_spec`.

The check uses the rather weak notion of register-compatibility as
the criterion. This means that any two reference types are considered
compatible, for example. It may be worth considering switching to
a stronger type compatibility relation at a later time.

### CODE REVIEW

File: src/SAWScript/Crucible/JVM/Builtins.hs
Comment (kquick) [inline src/SAWScript/Crucible/JVM/Builtins.hs:362]: Nothing wrong with this implementation, but could also be: ```registerCompatible = (==) `on` storageType```

Wondering though if `checkRegisterCompatibility` above should be changed to `return . registerCompatible` to eliminate duplicate implementations that risk diverging in the future.

**Review by kquick — COMMENTED** [review]
Looks good except for the concern about the duplication of register compatibility testing.

[Also not wild about seeing more `fail` but that's already the mechanism in use here, so that's a larger issue.]

**Comment by kquick** [comment]
One additional note is that it would be great if there were some tests for this in the test suite.  I suppose that depends on the maturity of the test suite and how easy it would be to add this, so I'll leave that up to your discretion.

File: src/SAWScript/Crucible/JVM/Builtins.hs
Comment (brianhuffman) [reply src/SAWScript/Crucible/JVM/Builtins.hs:362]: I should probaby just get rid of the monadic version completely; I was just too lazy to rewrite the use sites of the monadic one.

**Review by kquick — APPROVED** [review]
Oops, didn't mean to un-review.  Sometimes touchpads are too sensitive to things they think are a click but are just a touch.

**Comment by brianhuffman** [comment]
Adding a regression test is a good idea. I think I have some suitable files sitting around already, so it shouldn't be too much trouble.

File: src/SAWScript/Crucible/JVM/Builtins.hs
Comment (brianhuffman) [reply src/SAWScript/Crucible/JVM/Builtins.hs:357]: As of cf055b6b8ac45b8611f17847b165de4e8f79135c, the monadic `checkRegisterCompatibility` is gone.

**Comment by brianhuffman** [comment]
I got a little carried away and made a very comprehensive set of tests. Along the way, I discovered a bug that I fixed (
4c31c07) and I also noticed an awful lot of other well-formedness conditions that we should be checking, but don't yet (#938). So currently the test contains some proper tests, as well as a bunch of known failures.

**Comment by brianhuffman** [comment]
I think this is all ready to merge, pending CI tests.

**Review by kquick — APPROVED** [review]
Looks great, and I really appreciate the extensive suite of tests you created.

**Pull request merged**
